pkg/tsdb/influxdb/response_parser.go

```diff
@@ -21,48 +23,75 @@ func init() {
 	legendFormat = regexp.MustCompile(`\[\[([\@\/\w-]+)(\.[\@\/\w-]+)*\]\]*|\$\s*([\@\/\w-]+?)*`)
 }
 
-func (rp *ResponseParser) Parse(response *Response, query *Query) plugins.DataQueryResult {
+func (rp *ResponseParser) Parse(buf io.ReadCloser, query *Query) plugins.DataQueryResult {
 	var queryRes plugins.DataQueryResult
 
+	response, jsonErr := parseJSON(buf)
+	if jsonErr != nil {
+		queryRes.Error = jsonErr
+		return queryRes
+	}
+
+	if response.Error != "" {
+		queryRes.Error = fmt.Errorf(response.Error)
+		return queryRes
+	}
+
+	frames := data.Frames{}
 	for _, result := range response.Results {
-		queryRes.Series = append(queryRes.Series, rp.transformRows(result.Series, queryRes, query)...)
-		if result.Err != nil {
-			queryRes.Error = result.Err
+		frames = append(frames, transformRows(result.Series, query)...)
+		if result.Error != "" {
+			queryRes.Error = fmt.Errorf(result.Error)
 		}
 	}
+	queryRes.Dataframes = plugins.NewDecodedDataFrames(frames)
 
 	return queryRes
 }
 
-func (rp *ResponseParser) transformRows(rows []Row, queryResult plugins.DataQueryResult, query *Query) plugins.DataTimeSeriesSlice {
-	var result plugins.DataTimeSeriesSlice
+func parseJSON(buf io.ReadCloser) (Response, error) {
+	var response Response
+	dec := json.NewDecoder(buf)
+	dec.UseNumber()
+
+	err := dec.Decode(&response)
+	return response, err
+}
+
+func transformRows(rows []Row, query *Query) data.Frames {
+	frames := data.Frames{}
 	for _, row := range rows {
 		for columnIndex, column := range row.Columns {
 			if column == "time" {
 				continue
 			}
 
-			var points plugins.DataTimeSeriesPoints
+			var timeArray []time.Time
+			var valueArray []*float64
+
 			for _, valuePair := range row.Values {
-				point, err := rp.parseTimepoint(valuePair, columnIndex)
-				if err == nil {
-					points = append(points, point)
+				timestamp, timestampErr := parseTimestamp(valuePair[0])
+				// we only add this row if the timestamp is valid
+				if timestampErr == nil {
+					value := parseValue(valuePair[columnIndex])
+					timeArray = append(timeArray, timestamp)
+					valueArray = append(valueArray, value)
 				}
 			}
-			result = append(result, plugins.DataTimeSeries{
-				Name:   rp.formatSeriesName(row, column, query),
-				Points: points,
-				Tags:   row.Tags,
-			})
+			name := formatFrameName(row, column, query)
+
+			frames = append(frames, data.NewFrame(name,
+				data.NewField("time", nil, timeArray),
+				data.NewField("value", row.Tags, valueArray)))
```

Comment:
Tested this and realised that the way it is now, if you hover over a value, you see:

![image](https://user-images.githubusercontent.com/15115078/115200624-171eda00-a0fd-11eb-8ec0-a1836ce80b52.png)

I think you have to add the following suggestion.
```suggestion
				data.NewField("value", row.Tags, valueArray).SetConfig(&data.FieldConfig{DisplayNameFromDS: name})))
```

after this, it becomes like:

![image](https://user-images.githubusercontent.com/15115078/115200676-24d45f80-a0fd-11eb-8b9e-2ec1d7ffadf4.png)

Reply:
i added the `.SetConfig(&data.FieldConfig{DisplayNameFromDS: name})` , in slightly different format.

Reply:
could you try it now? it might fix the problems.